src/pymoca/tree.py

```diff
@@ -747,6 +775,494 @@ def _check_import_rules(
+        self,
+        orig_class: Union[ast.Class, ast.InstanceClass],
+        # Does InstanceElement need a fully_instantiated boolean flag?
+        modification_environment: ast.ClassModification,
+        parent: ast.Class,
+        partially=True,
+    ) -> ast.InstanceClass:
+        # Outline from spec 3.5 section 5.6.1 *Instantiation*:
+        # Definitions
+        #   - Element: Class, Component (Symbol in Pymoca), or Extends Clause
+        # 1. For element itself:
+        #   1. Create an instance of the class to be instantiated ("partially instantiated element")
+        #   2. Modifiers are merged for the element itself (but contained references are resolved during flattening)
+        #   3. Redeclare of element itself is done
+        # 2. For each element (Class or Component) in the local contents of the current element:
+        #   1. Apply step 1 to the element
+        #   2. Equations, algorithms, and annotations are copied into the component instance without merging
+        #      (but references in equations are resolved later during flattening)
+        # 3. For each element in the extends clauses of the current element:
+        #   1. Apply steps 1 and 2 to the element, replacing the extends clause with the extends instance
+        # 4. Lookup classes of extends and ensure it is identical to lookup result from step 3
+        # 5. Check that all children of the current element (including extends) with same name are identical
+        #    (error if not) and only keep one if so (to preserve function argument order)
+        # 6. Components are recursively instantiated
+
+        # 1.1. Partially instantiate the element itself and 1.2 merge modifiers
+        new_class = self._instantiate_partially(
+            orig_class,
+            modification_environment,
+            parent,
+        )
+
+        # 1.3. Redeclare of element itself is done
+        self._apply_class_redeclares(new_class, modification_environment)
+
+        # 2.1 Partially instantiate local classes and symbols
+        if isinstance(orig_class, ast.InstanceClass) and not orig_class.fully_instantiated:
+            from_class = new_class.ast_ref
+        else:
+            from_class = orig_class
+        for name, class_ in from_class.classes.items():
+            instance = self._instantiate_partially(
+                class_,
+                modification_environment,
+                new_class,
+            )
+            new_class.classes[name] = instance
+
+        for name, symbol in from_class.symbols.items():
+            instance = self._instantiate_partially(
+                symbol,
+                modification_environment,
+                new_class,
+            )
+            new_class.symbols[name] = instance
+
+        # 2.2 Copy local contents into the element itself
+        self._copy_class_contents(new_class, copy_extends=True, partially=partially)
+
+        # 3. Instantiate extends and 4. Check extends class lookup
```

Comment:
We really can't do that bit of the spec without tearing up or adding to our AST. Declarations would need to be kept in one (heterogenous) `list` or `dict` to preserve order instead of, or in addition to, our current `Class` attributes `classes`, `symbols`, `extends`, etc. So for now, we just do the `_instantiate_extends`.